src/agent/market_schedule.rs

```diff
@@ -0,0 +1,561 @@
+//! Holiday hours metadata parsing and evaluation logic
+
+use {
+    super::legacy_schedule::{
+        LegacySchedule,
+        MHKind,
+    },
+    anyhow::{
+        anyhow,
+        Result,
+    },
+    chrono::{
+        naive::NaiveTime,
+        DateTime,
+        Datelike,
+        Duration,
+        Utc,
+    },
+    chrono_tz::Tz,
+    proptest::{
+        arbitrary::any,
+        prop_compose,
+        proptest,
+    },
+    std::{
+        fmt::Display,
+        str::FromStr,
+    },
+    winnow::{
+        combinator::{
+            alt,
+            separated,
+            seq,
+        },
+        stream::ToUsize,
+        token::{
+            take,
+            take_till,
+        },
+        PResult,
+        Parser,
+    },
+};
+
+
+/// Helper time value representing 24:00:00 as 00:00:00 minus 1
+/// nanosecond (underflowing to 23:59:59.999(...) ). While chrono
+/// has this value internally exposed as NaiveTime::MAX, it is not
+/// exposed outside the crate.
+const MAX_TIME_INSTANT: NaiveTime = NaiveTime::MIN
+    .overflowing_sub_signed(Duration::nanoseconds(1))
+    .0;
+
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct MarketSchedule {
+    pub timezone:        Tz,
+    pub weekly_schedule: Vec<ScheduleDayKind>,
+    pub holidays:        Vec<HolidayDaySchedule>,
+}
+
+impl Default for MarketSchedule {
+    fn default() -> Self {
+        Self {
+            timezone:        Tz::UTC,
+            weekly_schedule: vec![ScheduleDayKind::Open; 7],
+            holidays:        vec![],
+        }
+    }
+}
+
+impl Display for MarketSchedule {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{};", self.timezone)?;
+        for (i, day) in self.weekly_schedule.iter().enumerate() {
+            write!(f, "{}", day)?;
+            if i < 6 {
+                write!(f, ",")?;
+            }
+        }
+        write!(f, ";")?;
+        for (i, holiday) in self.holidays.iter().enumerate() {
+            write!(f, "{}", holiday)?;
+            if i < self.holidays.len() - 1 {
+                write!(f, ",")?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl MarketSchedule {
+    pub fn can_publish_at(&self, when: &DateTime<Utc>) -> bool {
+        let when_local = when.with_timezone(&self.timezone);
+
+        let month = when_local.date_naive().month0() + 1;
+        let day = when_local.date_naive().day0() + 1;
+        let time = when_local.time();
+        let weekday = when_local.weekday().number_from_monday().to_usize();
+
+        for holiday in &self.holidays {
+            // Check if the day matches
+            if holiday.month == month && holiday.day == day {
+                return holiday.kind.can_publish_at(time);
+            }
+        }
+
+        self.weekly_schedule[weekday].can_publish_at(time)
```

Comment:
last Sunday we had an index out of bounds error here, and our pyth-agent panicked.
```shell
thread 'tokio-runtime-worker'
panicked at
index out of bounds:
src/agent/market_schedule.rs:108:29:
the len is 7 but the index is 7
note: run with 'RUST_BACKTRACE=1'
environment variable to display a backtrace
```

Reply:
Thanks for catching this. I pushed the fixed in another PR.